app/assets/javascripts/admin/order_cycles/controllers/incoming_controller.js.coffee

```diff
@@ -9,19 +9,23 @@ angular.module('admin.orderCycles').controller 'AdminOrderCycleIncomingCtrl', ($
     enterprise = $scope.enterprises[exchange.enterprise_id]
     return enterprise.numVariants if enterprise.numVariants?
 
-    $scope.loadExchangeProducts(exchange)
-    return unless enterprise.supplied_products?
+    enterprise.numVariants = 0
+    params = { exchange_id: exchange.id, enterprise_id: exchange.enterprise_id, order_cycle_id: $scope.order_cycle.id, incoming: true}
+    ExchangeProduct.countVariants params, (variants_count) ->
+      enterprise.numVariants = variants_count
+      $scope.setSelectAllVariantsCheckboxValue(exchange, enterprise.numVariants)
 
-    enterprise.numVariants = $scope.countVariants(enterprise.supplied_products)
-
-  $scope.countVariants = (products) ->
-    return 0 unless products
-
-    numVariants = 0
-    for product in products
-      numVariants += product.variants.length
-    numVariants
+    return enterprise.numVariants
 
   $scope.addSupplier = ($event) ->
     $event.preventDefault()
     OrderCycle.addSupplier $scope.new_supplier_id
+
+  # To select all variants we first need to load them all from the server
+  #
+  # This is only needed in Incoming exchanges as here we use supplied_products,
+  #   in Outgoing Exchanges the variants are loaded as part of the Exchange payload
+  $scope.selectAllVariants = (exchange, selected) ->
+    $scope.loadAllExchangeProducts(exchange).then ->
+      $scope.setExchangeVariants(exchange, $scope.suppliedVariants(exchange.enterprise_id), selected)
+      $scope.$apply()
```

Comment:
Having to manually call `$scope.apply()` is a code smell in Angular...

Reply:
yes, agree. this is the only way I managed to refresh the page after the load process finishes. angular is not triggering a re-render.
anyway, this is supposed to disappear as we implement #4487